xcm/src/v1/multilocation.rs

```diff
@@ -393,79 +740,74 @@ impl MultiLocation {
-			(5, MultiLocation::X8(_, _, _, _, _, ref mut a, ..)) => a,
-			(6, MultiLocation::X7(_, _, _, _, _, _, ref mut a)) => a,
-			(6, MultiLocation::X8(_, _, _, _, _, _, ref mut a, ..)) => a,
-			(7, MultiLocation::X8(_, _, _, _, _, _, _, ref mut a)) => a,
+			(0, Junctions::X1(ref mut a)) => a,
+			(0, Junctions::X2(ref mut a, ..)) => a,
+			(0, Junctions::X3(ref mut a, ..)) => a,
+			(0, Junctions::X4(ref mut a, ..)) => a,
+			(0, Junctions::X5(ref mut a, ..)) => a,
+			(0, Junctions::X6(ref mut a, ..)) => a,
+			(0, Junctions::X7(ref mut a, ..)) => a,
+			(0, Junctions::X8(ref mut a, ..)) => a,
+			(1, Junctions::X2(_, ref mut a)) => a,
+			(1, Junctions::X3(_, ref mut a, ..)) => a,
+			(1, Junctions::X4(_, ref mut a, ..)) => a,
+			(1, Junctions::X5(_, ref mut a, ..)) => a,
+			(1, Junctions::X6(_, ref mut a, ..)) => a,
+			(1, Junctions::X7(_, ref mut a, ..)) => a,
+			(1, Junctions::X8(_, ref mut a, ..)) => a,
+			(2, Junctions::X3(_, _, ref mut a)) => a,
+			(2, Junctions::X4(_, _, ref mut a, ..)) => a,
+			(2, Junctions::X5(_, _, ref mut a, ..)) => a,
+			(2, Junctions::X6(_, _, ref mut a, ..)) => a,
+			(2, Junctions::X7(_, _, ref mut a, ..)) => a,
+			(2, Junctions::X8(_, _, ref mut a, ..)) => a,
+			(3, Junctions::X4(_, _, _, ref mut a)) => a,
+			(3, Junctions::X5(_, _, _, ref mut a, ..)) => a,
+			(3, Junctions::X6(_, _, _, ref mut a, ..)) => a,
+			(3, Junctions::X7(_, _, _, ref mut a, ..)) => a,
+			(3, Junctions::X8(_, _, _, ref mut a, ..)) => a,
+			(4, Junctions::X5(_, _, _, _, ref mut a)) => a,
+			(4, Junctions::X6(_, _, _, _, ref mut a, ..)) => a,
+			(4, Junctions::X7(_, _, _, _, ref mut a, ..)) => a,
+			(4, Junctions::X8(_, _, _, _, ref mut a, ..)) => a,
+			(5, Junctions::X6(_, _, _, _, _, ref mut a)) => a,
+			(5, Junctions::X7(_, _, _, _, _, ref mut a, ..)) => a,
+			(5, Junctions::X8(_, _, _, _, _, ref mut a, ..)) => a,
+			(6, Junctions::X7(_, _, _, _, _, _, ref mut a)) => a,
+			(6, Junctions::X8(_, _, _, _, _, _, ref mut a, ..)) => a,
+			(7, Junctions::X8(_, _, _, _, _, _, _, ref mut a)) => a,
 			_ => return None,
 		})
 	}
 
 	/// Returns a reference iterator over the junctions.
-	pub fn iter(&self) -> MultiLocationRefIterator {
-		MultiLocationRefIterator(&self, 0)
+	pub fn iter(&self) -> JunctionsRefIterator {
+		JunctionsRefIterator(&self, 0)
 	}
 
 	/// Returns a reference iterator over the junctions in reverse.
-	pub fn iter_rev(&self) -> MultiLocationReverseRefIterator {
-		MultiLocationReverseRefIterator(&self, 0)
-	}
-
-	/// Consumes `self` and returns an iterator over the junctions.
-	pub fn into_iter(self) -> MultiLocationIterator {
-		MultiLocationIterator(self)
+	pub fn iter_rev(&self) -> JunctionsReverseRefIterator {
+		JunctionsReverseRefIterator(&self, 0)
 	}
 
 	/// Consumes `self` and returns an iterator over the junctions in reverse.
-	pub fn into_iter_rev(self) -> MultiLocationReverseIterator {
-		MultiLocationReverseIterator(self)
+	pub fn into_iter_rev(self) -> JunctionsReverseIterator {
```

Comment:
again, why not do a normal iterator and just call `.rev()`?

Reply:
I can imagine that maybe this is more efficient, since `iter().rev()` will need to traverse the normal iterator to obtain the reverse.

Reply:
We could, I'm not exactly sure why this exists honestly, since implementing iterators properly involves implementing the `IntoIterator` trait.

Which does give a a good idea -- perhaps instead of making an `into_iter_rev` method, we can implement a custom `rev` method in the `impl Iterator for JunctionsIterator` trait impl and make use of `JunctionsReverseIterator` there...

Reply:
Nope, impossible, because the `rev()` method returns https://doc.rust-lang.org/stable/std/iter/struct.Rev.html and doesn't return `impl Iterator` :disappointed: .

Reply:
Oh, but what we can do is implement `DoubleEndedIterator` on `Junctions` and remove `JunctionsReverseIterator` altogether and everything will work!